Update android.md

Camerash opened the pull request:
-Added logout and change username/email sessions
-Refactor to more clear sessions
-Align with JS docs
-Updated all onFail callback method to return a Error object (instead of the old String reason)

Camerash committed:
Update android.md

-Added logout and change username/email sessions
-Refactor to more clear sessions
-Align with JS docs
-Updated all onFail callback method to return a Error object (instead of the old String reason)

oursky-travis commented:
@Camerash, thanks for your PR! By analyzing the history of the files in this pull request, we identified @bensonby to be a potential reviewer.

bensonby commented on auth/basics/android.md:
```diff
@@ -46,19 +53,45 @@ skygear.signupWithEmail(email, password, new AuthResponseHandler() {
+### Anonymous user
 
-The following shows how to log in a user with username and password.
+Skygear support users to sign up anonymously. However, they can only read data from the public database, but cannot perform operations such as saving data.
```
- Grammar: "Skygear support_s_ anonymous sign-up"
- I believe anonymous user can also save records and do anything as if the same as a normal user, could anyone confirm? If not the JS doc has to be modified as well

chpapa replied:
yes I think so, could @ben181231 confirm?

ben181231 replied:
Yes. Anonymous user is just a normal user. He can update record, read record while the ACL allows.

Camerash replied:
Then I believe the JS docs of User Auth Basics is needed to be update , because it stated that anonymous user can only read records, but not write.
https://docs.skygear.io/guides/auth/basics/js/#signing-up-logging-in-logging-out

tensiuyan replied:
@Camerash Thanks for spotting that. I understand what you mean. I guess the explanation in the JS guide about anonymous user is not clear enough. Let me try update it.

chpapa replied:
@tensiuyan @Camerash I read the JS guide again and it is actually pretty clear about it. I think we might want to change this line to make it clearer: 

> Without being authenticated, a user can read data from the public database but cannot perform most of the other operations, including saving data into the database.

to this

> Without signing in, a client can only read data from the public database but cannot perform most of the other operations, including saving data into the database.

Just don't mix up about "authenticated" and "anonymous" user.

bensonby commented on auth/basics/android.md:
````diff
@@ -46,19 +53,45 @@ skygear.signupWithEmail(email, password, new AuthResponseHandler() {
 
 ```java
-Container skygear = Container.defaultContainer(this);
+skygear.signupAninymously(new AuthResponseHandler() {
````
Spelling: `Aninymously` -> `Anonymously`

bensonby commented on auth/basics/android.md:
```diff
@@ -46,19 +53,45 @@ skygear.signupWithEmail(email, password, new AuthResponseHandler() {
+skygear.signupAninymously(new AuthResponseHandler() {
+    @Override
+    public void onAuthSuccess(User user) {
+        Log.i("Skygear Signup", "onAuthSuccess: Got token: " + user.accessToken);
```
I believe it will be better to explain what an `access token` is, perhaps somewhere above as an "Overview" of authentication

bensonby commented on auth/basics/android.md:
```diff
@@ -46,19 +53,45 @@ skygear.signupWithEmail(email, password, new AuthResponseHandler() {
+
+Each anonymous user has a unique ID that behaves exactly the same way as normal. As expected, an anonymous user does not have username, email nor password.
+
+Because of this behaviour, once the token of an anonymous user is lost, so will the account.
```
To me, the word "lost" is a little ambiguous here. The user, the account, and the saved records will persist, just that the user cannot log in as that anonymous user again

bensonby commented on auth/basics/android.md:
```diff
@@ -46,19 +53,45 @@ skygear.signupWithEmail(email, password, new AuthResponseHandler() {
+
+Logging in behave much the same way as signing up a user.
+
+However, if user try to sign up with an exsisting username / email, the error `ERROR.CODE.DUPLICATED` will be returned in the `onAuthFail` method, indicating the username / email is already registered.
```
- Will be better to show an example using `ERROR.CODE.DUPLICATED`, but will this sign-up error handling part mentioned in the "Sign-up" part instead of logging in?
- As a best practice we usually keep line length within 80 characters for easier editing

bensonby commented on auth/basics/android.md:
```diff
@@ -123,8 +171,48 @@ skygear.whoami(new AuthResponseHandler() {
+
+To change a user's username and email, you can use the `skygear.saveUser()` method by providing the user ID and the new username and/or the new email.
+
+Every user can only edit his/her information
```
Grammar: "A user" or "Each user"

bensonby commented on auth/basics/android.md:
```diff
@@ -123,8 +171,48 @@ skygear.whoami(new AuthResponseHandler() {
+To change a user's username and email, you can use the `skygear.saveUser()` method by providing the user ID and the new username and/or the new email.
+
+Every user can only edit his/her information
+While only users with admin privilege can change information of other users.
```
The word "admin privilege" will be confusing because this is not mentioned/explained in other places in this document

bensonby commented on auth/basics/android.md:
```diff
@@ -123,8 +171,48 @@ skygear.whoami(new AuthResponseHandler() {
+User newUser = new User(currentUser.userId, 
+                        currentUser.accessToken, 
+                        "your-new-username",
+                        "your-new-email@hello.com"); 
```
Can I change only the email but not username?

bensonby reviewed (commented):
Thanks @Camerash for revising the page. I haven't tested the code examples - I only read through it once and made some comments on this PR. Given that I won't be responsible for the doc, it will be better to assign to somebody else for further review/approval.

bensonby commented:
With the comment https://github.com/SkygearIO/guides/pull/68#pullrequestreview-35539715, I will assign this to @ben181231 first for now

ben181231 commented:
Except [the comment](https://github.com/SkygearIO/guides/pull/68#discussion_r114242056) about the permissions of anonymous user, others looks good to me.

Camerash committed:
Update android.md

-Fixed various spelling and grammar mistakes
-Revised word usage
-New Overview part

tensiuyan committed:
fix typos

tensiuyan commented on auth/basics/android.md:
```diff
@@ -230,9 +249,9 @@ skygear.saveUser(newUser, new UserSaveResponseHandler() {
+If you only want to change one of the user fields, just replace the new username/email with `currentUser.getUsername()`/`currentUser.getEmail()`.
 
-### Change the password of a user
+## Updating the password of a user
```
@Camerash I saw there's a section in JS about updating the password of a user https://docs.skygear.io/guides/auth/basics/js/#changing-the-username-and-email-of-a-user

Do we have this API in Android too?

tensiuyan reviewed (commented):
@Camerash thanks for the update! I have briefly fixed some typos and added some explanations. Let me know what do you think about it. 

BTW, I was cross checking this guide with the JS one. I found that there are 2 sections missing:

- Observing user changes
- Updating password

May I know if it is the case that we haven't supported them in the Android SDK?

ben181231 commented:
@Camerash @tensiuyan Any updates for this PR ?

Camerash commented:
Sorry for the lack of updates as I just finished my final exam.

I have checked the Android SDK and it seems it doesn't have support over
changing a user's password and observing user change (or I am just missing
it).

On Jun 5, 2017 10:40, "Ben Lei" <notifications@github.com> wrote:

@Camerash <https://github.com/camerash> @tensiuyan
<https://github.com/tensiuyan> Any updates for this PR ?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/SkygearIO/guides/pull/68#issuecomment-306090114>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AVIz1wtbhNHM5trmATkLTY54C3lP-x6nks5sA2qlgaJpZM4NMkx7>
.

Steven-Chan commented:
This PR is a bit out dated, I have create a new PR https://github.com/SkygearIO/guides/pull/83 from this branch, rebased on the current master.
